### ISSUE TITLE
[clang] [windows] [address-sanitizer] stack overflow when test is failing

### ISSUE THREAD

**Issue by hia3** [catchorg/Catch2#898]
## Description
Catch enters infinite recursion trying to print error message when compiled with clang -faddress-sanitize under Windows

### Steps to reproduce
0. Install VS 2015 and [clang](http://llvm.org/pre-releases/win-snapshots/LLVM-5.0.0-r300231-win64.exe)
1. Put `catch.hpp` and `test.bat` from [test.zip](https://github.com/philsquared/Catch/files/964783/test.zip) in one directory, start `test.bat`
2. Resulting executable, instead of reporting an error, will output:
<pre>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

</pre>
and then it will crash.

**Comment by kirbyfan64** [comment]
FWIW this works normally under Linux.

Does `-fsanitize=address` work normally for you outside of Catch? (e.g. if you build a large program with it enabled, does it work correctly?) What if you use normal Clang:

```
clang++ -fsanitize=address -o test.exe test.cpp
```

**Comment by hia3** [comment]
Compiling with 

    clang++ -fsanitize=address -o test.exe test.cpp

gives me 

<pre>
==748==AddressSanitizer CHECK failed: D:\src\llvm_package_300231\llvm\projects\compiler-rt\lib\asan\asan_rtl.cc:505 "((!asan_init_is_running && "ASan init calls itself!")) != (0)" (0x0, 0x0)
==748==AddressSanitizer CHECK failed: D:\src\llvm_package_300231\llvm\projects\compiler-rt\lib\asan\asan_poisoning.cc:37 "((AddrIsInMem(addr))) != (0)" (0x0, 0x0)
</pre>
 
even if 'test.cpp' is just an empty `main`. This is why I've added `/DEBUG` when linking.

With `clang_rt.asan_dynamic-x86_64.lib` linking fails:
<pre>
test.obj : error LNK2019: unresolved external symbol __asan_shadow_memory_dynamic_address referenced in function "struct Catch::IResultCapture & __cdecl Catch::getResultCapture(void)" (?getResultCapture@Catch@@YAAEAUIResultCapture@1@XZ)
test.obj : error LNK2019: unresolved external symbol __asan_option_detect_stack_use_after_return referenced in function "struct Catch::IResultCapture & __cdecl Catch::getResultCapture(void)" (?getResultCapture@Catch@@YAAEAUIResultCapture@1@XZ)
test.exe : fatal error LNK1120: 2 unresolved externals
</pre>

**Comment by horenmar** [comment]
Is ASan supported under Windows? Last time I checked, it was "this will very likely work for C, C++ is very much beta".

**Comment by ekilmer** [comment]
Hello. Now that Address Sanitizer is [Generally Available](https://devblogs.microsoft.com/cppblog/address-sanitizer-for-msvc-now-generally-available/) for MSVC, I've noticed that I'm unable to run my tests due to this issue--they get stuck and won't complete.

After some digging, I found a relevant open issue on Microsoft's tracker with a [potential workaround](https://developercommunity.visualstudio.com/t/Asan-problems-when-manual-exception-hand/1242524#T-N1295191) that would consist of the following (naive) change to Catch here https://github.com/catchorg/Catch2/blob/b025a007b924cc16da84c07a5d35118695cb9bcc/include/internal/catch_fatal_condition.cpp#L64 and here for the single include header: https://github.com/catchorg/Catch2/blob/b025a007b924cc16da84c07a5d35118695cb9bcc/single_include/catch2/catch.hpp#L10787

```diff
- exceptionHandlerHandle = AddVectoredExceptionHandler(1, handleVectoredException);
+ exceptionHandlerHandle = AddVectoredExceptionHandler(0, handleVectoredException);
```


I've tested this change locally with a modified (read: updated) version of the batch script in the OP with a fully updated Visual Studio (v16.9.1) on latest Catch (v2.13.4), and it seems to run to completion! 
```bat
echo #define CATCH_CONFIG_MAIN > test.cpp
echo #include "catch.hpp" >> test.cpp
echo TEST_CASE( "t", "[t]" ) { REQUIRE( 2 == 1 ); } >> test.cpp

call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64
echo Please wait...
cl /EHsc test.cpp -fsanitize=address -c
link /DEBUG test.obj

start test.exe
```

However, I don't understand the consequences of this change and how it could affect the rest of Catch's logic and exception handling, so hopefully this helps one of the maintainers or someone else more familiar with the code base and C++ in general 🙂

**Comment by reder2000** [comment]
This is related to the way asan is implemented on x64 msvc. See here [doc](https://devblogs.microsoft.com/cppblog/asan-for-windows-x64-and-debug-build-support/).
Catch2 is catching tons of 0xc0000005 (FCE - first chance exceptions) and gets confused.
Your fix works but gives up on FCEs.

catch2+asan+msvc is working on x86 unmodified (but nobody builds for x86 anymore)

Maybe catch2 could offer an option to give up on first chance exception or do it automatically on x64 with
```#  if __SANITIZE_ADDRESS__
... ```

Edit: with catch2 moving as a library, this way to fix depending on asan would not work.

**Comment by unixwitch** [comment]
as a workaround, setting `CATCH_CONFIG_NO_WINDOWS_SEH` fixed this issue for me (tested with 2.13.6).
